### PR TITLE
Multicluster glooctl commands and failover helm values

### PR DESCRIPTION
**Pull request by ashleywang1**
# Description

This is a collection of glooctl and Failover changes necessary to make Gloo work with Gloo Federation.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo [contribution guide](https://docs.solo.io/gloo/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/3369
resolves https://github.com/solo-io/gloo/issues/3338

### CODE REVIEW

**Comment by ashleywang1** [comment]
Planning to do envoy v2->v3 update with this PR: #3343

File: changelog/v1.5.0-beta9/gloo-federation-branch.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      Add the following commands:
+      - `glooctl install federation` - install Gloo Federation on Kubernetes
+      - `glooctl uninstall federation` - uninstall Gloo Federation
+    issueLink: https://github.com/solo-io/gloo/issues/3369
+  - type: HELM
+    description: |
+      Add the following Helm values, available for Gloo Enterprise only:
+      - `gatewayProxies.NAME.failover.enabled`: Configure this proxy for failover.
+      Set to false by default.
+      - `gatewayProxies.NAME.failover.port`: Port to use for failover Gateway Bind port,
```
Comment (kdorosh) [inline changelog/v1.5.0-beta9/gloo-federation-branch.yaml:20]: nit, capitalization here is a bit off "Bind"

File: install/helm/gloo/templates/19-gloo-mtls-configmap.yaml [hunk 2]
```diff
@@ -16,14 +16,43 @@ data:
...
         - filters:
           - name: envoy.tcp_proxy
             config:
               stat_prefix: ingress_tcp
-              cluster: gloo_cluster
+              cluster: gloo_xds_cluster
+          transport_socket:
+            name: envoy.transport_sockets.tls
+            typed_config:
+              "@type": type.googleapis.com/envoy.api.v2.auth.DownstreamTlsContext
```
Comment (kdorosh) [inline install/helm/gloo/templates/19-gloo-mtls-configmap.yaml:30]: new stuff should use v3 apis, `type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext`

File: install/helm/gloo/templates/9-gateway-proxy-configmap.yaml [hunk 3]
```diff
@@ -179,6 +179,45 @@ data:
...
+                    port_value: {{ $.Values.gloo.deployment.restXdsPort }}
+        upstream_connection_options:
+          tcp_keepalive: {}
+        type: STRICT_DNS
+        respect_dns_ttl: true
+{{- if $global.glooMtls.enabled }}
+        transport_socket:
+          name: envoy.transport_sockets.tls
+          typed_config:
+            "@type": type.googleapis.com/envoy.api.v2.auth.UpstreamTlsContext
```
Comment (kdorosh) [inline install/helm/gloo/templates/9-gateway-proxy-configmap.yaml:203]: lets use v3, `"@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext`

File: install/helm/gloo/templates/19-gloo-mtls-configmap.yaml [hunk 4]
```diff
@@ -16,14 +16,43 @@ data:
...
+                      api_type: GRPC
+                      grpc_services:
+                      - envoy_grpc:
+                          cluster_name: gloo_sds
+      - name: envoy_rest_xds_mtls_listener
+        address: { socket_address: { address: "0.0.0.0", port_value: {{ .Values.gloo.deployment.restXdsPort }} } }
+        filter_chains:
+        - filters:
+          - name: envoy.tcp_proxy
+            config:
```
Comment (kdorosh) [inline install/helm/gloo/templates/19-gloo-mtls-configmap.yaml:53]: lets use v3
```
- name: envoy.filters.network.tcp_proxy
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
```

File: hack/find_latest_enterprise_version.go [hunk 5]
```diff
@@ -37,7 +37,7 @@ func main() {
 		log.Fatal(err)
 	}
 	defer f.Close()
-	enterpriseVersion, err := version.GetLatestEnterpriseVersionWithMaxVersion(true, maxGlooEVersion)
+	enterpriseVersion, err := version.GetLatestHelmChartVersionWithMaxVersion(true, maxGlooEVersion)
```
Comment (kdorosh) [inline hack/find_latest_enterprise_version.go:40]: does this even compile? looks like it should be
```golang
GetLatestHelmChartVersionWithMaxVersion(EnterpriseHelmRepoIndex, GlooEE, true, maxGlooEVersion)
```

File: projects/gloo/api/v1/settings.proto [hunk 6]
```diff
@@ -390,6 +390,10 @@ message GlooOptions {
     // Set this option to specify the default max program size for regexes. If not specified, 
     // defaults to 100.
     google.protobuf.UInt32Value regex_max_program_size = 10;
+
+    // (Enterprise Only): Where the `gloo` REST xDS server should bind (should not need configuration by user).
```
Comment (kdorosh) [inline projects/gloo/api/v1/settings.proto:394]: "should not need config by user" --> this is a bit confusing on the comment for user config. what exactly should not need user config? why is that relevant to this config below?

File: projects/gloo/cli/pkg/cmd/federation/list/root.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	clusterSecretType = "solo.io/kubeconfig"
+)
+
+func RootCmd(opts *options.Options) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   constants.CLUSTER_LIST_COMMAND.Use,
+		Short: constants.CLUSTER_LIST_COMMAND.Short,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			secretClient := helpers.MustKubeClient().CoreV1().Secrets(opts.Cluster.FederationNamespace)
+			secrets, err := secretClient.List(metav1.ListOptions{})
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/federation/list/root.go:26]: lets get a namespaced client if we can so this doesn't require cluster scoped credentials

File: projects/gloo/cli/pkg/cmd/federation/register/root.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return Register(opts)
+		},
+	}
+
+	pflags := cmd.PersistentFlags()
+	flagutils.AddClusterFlags(pflags, &opts.Cluster)
+	flagutils.AddRegisterFlags(pflags, &opts.Cluster.Register)
+	// this flag is mainly for demo, testing, and debugging purposes
+	pflags.Lookup(flagutils.LocalClusterDomainOverride).Hidden = true
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/federation/register/root.go:27]: TIL hidden cobra flags, very cool

File: projects/gloo/cli/pkg/cmd/federation/register/register.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	},
+	{
+		Verbs:     []string{"get", "list", "watch"},
+		APIGroups: []string{""},
+		Resources: []string{"pods", "nodes", "services"},
+	},
+}
+
+func Register(opts *options.Options) error {
+	ctx := context.TODO()
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/federation/register/register.go:31]: should only use TODO context in testing. if we need a new context, we should make a background context

File: projects/gloo/cli/pkg/cmd/federation/root.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/solo-io/gloo/projects/gloo/cli/pkg/cmd/federation/unregister"
+	"github.com/solo-io/gloo/projects/gloo/cli/pkg/cmd/options"
+	"github.com/solo-io/gloo/projects/gloo/cli/pkg/constants"
+	"github.com/solo-io/gloo/projects/gloo/cli/pkg/prerun"
+
+	"github.com/solo-io/go-utils/cliutils"
+	"github.com/spf13/cobra"
+)
+
+var EmptyClusterError = eris.New("please provide a subcommand")
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/federation/root.go:16]: error name and error message don't seem aligned at first glance

File: projects/gloo/cli/pkg/cmd/federation/unregister/root.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+var EmptyClusterNameError = eris.New("please provide a cluster name to unregister")
+
+func RootCmd(opts *options.Options, optionsFunc ...cliutils.OptionsFunc) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   constants.CLUSTER_UNREGISTER_COMMAND.Use,
+		Short: constants.CLUSTER_UNREGISTER_COMMAND.Short,
+		Long:  constants.CLUSTER_UNREGISTER_COMMAND.Long,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			secretClient := helpers.MustSecretClient()
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/federation/unregister/root.go:25]: same concern about getting a namespaced secret client

File: projects/gloo/cli/pkg/cmd/install/installer.go [hunk 12]
```diff
@@ -101,11 +112,14 @@ func (i *installer) Install(installerConfig *InstallerConfig) error {
 	}
 
 	// determine if it's an enterprise chart by checking if has gloo as a dependency
-	installerConfig.Enterprise = false
-	for _, dependency := range chartObj.Dependencies() {
-		if dependency.Metadata.Name == constants.GlooReleaseName {
-			installerConfig.Enterprise = true
-			break
+	// if so, overwrite the installation mode to Enterprise
+	if installerConfig.Mode != Federation {
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/install/installer.go:116]: we don't want this check. this check allows users to provide a helm chart install via -f flag and automatically determine if we have an enterprise chart even if we're installing from a different sub command.

perhaps better UX here would be to loudly fail rather than silently fix the helm values we pass through

this change allows users to `glooctl install gateway -f <helm glooe chart` and have this succeed. with the check above we could not do the same with federation

File: hack/find_latest_enterprise_version.go (hunk 5 above)
Comment (ashleywang1) [reply hack/find_latest_enterprise_version.go:40]: Yep! It was actually only ever used during this makefile target.
$(OUTPUT_DIR)/gloo-enterprise-version:
	GO111MODULE=on go run hack/find_latest_enterprise_version.go

Comment (ashleywang1) [reply hack/find_latest_enterprise_version.go:40]: Changed to enterpriseVersion, err := version.GetLatestEnterpriseVersion(true)

File: projects/gloo/api/v1/settings.proto (hunk 6 above)
Comment (ashleywang1) [reply projects/gloo/api/v1/settings.proto:394]: Ah, it was copied from the comments for 'xds_bind_addr':
// Where the `gloo` xDS server should bind (should not need configuration by user). Defaults to `0.0.0.0:9977`

What would you think of removing  the "(should not need configuration by user)" part? What it's saying is that users really shouldn't be setting this value, and they should use the default.

Comment (kdorosh) [reply projects/gloo/api/v1/settings.proto:394]: Got it. As a user, it is a bit confusing to have a config and then tell me not to use it without context. Perhaps drop a note in there about how Gloo Fed will make use of it?

File: install/helm/gloo/templates/19-gloo-mtls-configmap.yaml (hunk 2 above)
Comment (ashleywang1) [reply install/helm/gloo/templates/19-gloo-mtls-configmap.yaml:30]: Fixed both here and on l:59

File: install/helm/gloo/templates/9-gateway-proxy-configmap.yaml (hunk 3 above)
Comment (ashleywang1) [reply install/helm/gloo/templates/9-gateway-proxy-configmap.yaml:203]: Fixed both here and on l:152

File: projects/gloo/cli/pkg/cmd/federation/list/root.go (hunk 7 above)
Comment (kdorosh) [reply projects/gloo/cli/pkg/cmd/federation/list/root.go:26]: actually this is the namespaced client, the real kube client (rather than our secret client).

you may still want to use our client though

File: projects/gloo/cli/pkg/cmd/federation/unregister/root.go (hunk 11 above)
Comment (ashleywang1) [reply projects/gloo/cli/pkg/cmd/federation/unregister/root.go:25]: Was able to fix it here!

File: projects/gloo/cli/pkg/cmd/federation/list/root.go (hunk 7 above)
Comment (ashleywang1) [reply projects/gloo/cli/pkg/cmd/federation/list/root.go:26]: I'm able to use the Gloo Client in the 'unregister' file, but here, for some reason, calling `secrets, err := secretClient.List(opts.Cluster.FederationNamespace, clients.ListOpts{})` doesn't return a list of secrets.

File: projects/gloo/api/v1/settings.proto (hunk 6 above)
Comment (ashleywang1) [reply projects/gloo/api/v1/settings.proto:394]: // (Enterprise and Gloo Federation Only): Where the `gloo` REST xDS server should bind.
?

Comment (kdorosh) [reply projects/gloo/api/v1/settings.proto]: Yeah that's much more clear
